### ISSUE TITLE
C++: Missing variable access target when target is a coroutine parameter

### ISSUE THREAD

**Issue by jcmiao** [github/codeql#12490]
**Description of the issue**

Everything worked fine when the pointer variable p was locally defined, but when I adjusted it to function arguments, I failed to get the variable through `getTarget()`

The code is as follows:
```
std::pair<int, int>* gp = nullptr;
CoroutineTaskVoid co_func(std::pair<int, int>* p)
{
    // it is ok for p2
    std::pair<int, int>* p2 = new std::pair<int, int>(3,3);
    gp = p;
    std::cout << "p->first: " << p->first << std::endl;
    co_await std::suspend_always{};
    (*p).second += 10;
    {
        // safe use
        int *p = new int(0);
        *p = 10;
    }
    std::cout << "p->second: " << p->second << std::endl;
    int n = 200;
    std::cout << "n is " << n << std::endl;
    co_await std::suspend_always{};
    // unsafe
    p2->first = 4;
    p->first = 2;  // PointerFieldAccess.(VariableAccess).getTarget() failed
    p = gp;
    p->first *= 2;
    p->second *= 4;
    co_return;
}
```

**Comment by smowton** [comment]
I elaborated your example as follows and thus established this is specifically a coroutines bug:

```cpp
#include <iostream>
#include <coroutine>

struct promise;
 
struct coroutine : std::coroutine_handle<promise>
{
    using promise_type = struct promise;
};
 
struct promise
{
    coroutine get_return_object() { return {coroutine::from_promise(*this)}; }
    std::suspend_always initial_suspend() noexcept { return {}; }
    std::suspend_always final_suspend() noexcept { return {}; }
    void return_void() {}
    void unhandled_exception() {}
};

std::pair<int, int>* gp = nullptr;
coroutine co_func(std::pair<int, int>* p)
{
    // it is ok for p2
    std::pair<int, int>* p2 = new std::pair<int, int>(3,3);
    gp = p;
    std::cout << "p->first: " << p->first << std::endl;
    co_await std::suspend_always{};
    (*p).second += 10;
    {
        // safe use
        int *p = new int(0);
        *p = 10;
    }
    std::cout << "p->second: " << p->second << std::endl;
    int n = 200;
    std::cout << "n is " << n << std::endl;
    co_await std::suspend_always{};
    // unsafe
    p2->first = 4;
    p->first = 2;  // PointerFieldAccess.(VariableAccess).getTarget() failed
    p = gp;
    p->first *= 2;
    p->second *= 4;
    co_return;
}

void nonco_func(std::pair<int, int>* p)
{
    // it is ok for p2
    std::pair<int, int>* p2 = new std::pair<int, int>(3,3);
    gp = p;
    std::cout << "p->first: " << p->first << std::endl;
    (*p).second += 10;
    {
        // safe use
        int *p = new int(0);
        *p = 10;
    }
    std::cout << "p->second: " << p->second << std::endl;
    int n = 200;
    std::cout << "n is " << n << std::endl;
    // unsafe
    p2->first = 4;
    p->first = 2;  // PointerFieldAccess.(VariableAccess).getTarget() failed
    p = gp;
    p->first *= 2;
    p->second *= 4;
}
```

QL:

```ql
import cpp

from PointerFieldAccess p
select p, p.getQualifier().(VariableAccess).getTarget()
```

**Comment by jketema** [comment]
Hi @jcmiao,

Thanks for your question. It seems that the data we put in the database for coroutines is incomplete, which causes `getTarget()` to not report a result. I've filed an internal issues for this to track this problem. Note that C++20 support (including coroutines) is currently in beta.

**Comment by jcmiao** [comment]
> Hi @jcmiao,
> 
> Thanks for your question. It seems that the data we put in the database for coroutines is incomplete, which causes `getTarget()` to not report a result. I've filed an internal issues for this to track this problem. Note that C++20 support (including coroutines) is currently in beta.

Ok, thanksfor your answer, I will keep an eye on the following updates

**Comment by jketema** [comment]
This should be fixed in CodeQL 2.17.1. Note that the result of `PointerFieldAccess.(VariableAccess).getTarget()` will be a local variable and not the - possibly expected - coroutine argument. The value of the argument will have been assigned to this local variable.

Closing this for now.
